### DISCUSSION TITLE
What new project types should projen have?

### DISCUSSION THREAD

**Discussion by mrgrain** [projen/projen#2214] [General]
Please share your ideas for new project types to be supported in projen. And don't forget to upvote your favorites! 🚀 

This might inform contributors what to work on next.

**Comment by mrgrain** [comment]
[Ansible project](https://github.com/projen/projen/issues/1793) (suggested by @nthienan)

**Comment by mrgrain** [comment]
[octoherd-script-ts](https://github.com/projen/projen/issues/2841#issue-1831665741) (suggested by @ahammond)

**Comment by mrgrain** [comment]
**cdktf-app-ts** (and maybe other jsii supported languages) (suggested by @valebedu)

**Comment by mrgrain** [comment]
**go** (suggested by @valebedu)

go does't require a lot of boilerplate code but a nice looking structure (cmd, docs etc), .golangciyml, .goreleaser.yml go.mod (optionnaly go.work) and a Makefile to run build test and run commands

**Reply by v41e** [reply]
I've worked on a `GoProject` and `GoWorkspace` (component) internally. I could propose a public package for this, but I think it might be interesting to have Go support directly in projen, as it is already used by the publisher and is a language supported by JSII.

I've also worked on an `HugoProject`, but I'm not sure if adding it to projen is justified; it could live in another repository.

**Reply by v41e** [reply]
@mrgrain any opinion regarding this?

**Reply by mrgrain** [reply]
Realistically we can't currently support any new project types in projen core. So any new project should go into a separate community maintained package.

I'm committed to help resolve any blockers in upstream if there are any.

**Reply by v41e** [reply]
Understood, thank you for the clarification.

Just to make sure I fully understand the constraint: is the decision mainly driven by a desire to limit scope and stabilize projen towards a 1.0 release, or is it more about long-term maintenance capacity for new project types?

In any case, I’m fine with exploring a community-maintained package and will follow that path if that’s the preferred approach and maybe link it from https://github.com/projen/awesome

**Reply by mrgrain** [reply]
> Just to make sure I fully understand the constraint: is the decision mainly driven by a desire to limit scope and stabilize projen towards a 1.0 release, or is it more about long-term maintenance capacity for new project types?

Neither really. The main reason is that we maintainers are primarily TypeScript users. Outside of that ecosystem is difficult for us to provide value, reviewing comes with a higher cost and diverts resources away from other PRs. We already see this with Python. So at this point, we'd rather see the community owning new base project types until we have established a new maintainer model that's more suitable. It's something we're thinking about, but will take time.

**Reply by ahammond** [reply]
> I've also worked on an `HugoProject`, but I'm not sure if adding it to projen is justified; it could live in another repository.

I would be interested in your Hugo work!

**Reply by v41e** [reply]
Thanks @mrgrain, that makes sense.

Is there a recommended pattern for proposing and structuring projects and components outside of `projen` core?

For example, would you prefer:
- multiple focused packages (e.g. one for Go-related projects such as `GoProject`, `GoWorkspace`, releaser/config helpers, another for OpenAPI-related projects), or
- a single package aggregating multiple project types, similar to how `projen` itself is structured, with a single dependency surface?

@ahammond happy to hear that! I’ll publish the Go and Hugo-related work soon so you can take a look.

**Reply by hoegertn** [reply]
I think having structured, focused on a project group makes sense. Like everything Go related, everything API related. That makes it easier to have maintainers that really know that particular framework or language.

**Reply by mrgrain** [reply]
I would `GoProject` and `GoWorkspace` into the same package, but I would put `HugoProject` into a separate one.

**Comment by blntrsz** [comment]
Hi @mrgrain I'd loved to see vite project, it could be a general project that can be extended into React vite project or Vue vite project.
It could cover most frontend needs as I see more vite applications then CRA apps nowadays

**Reply by mrgrain** [reply]
Any interest in building this out? Probably best to start as a separate package.

**Reply by blntrsz** [reply]
Yeah, would be interesting, I'll give it a go

**Reply by mbasadi** [reply]
Any progress on this? @blntrsz

**Reply by automartin5000** [reply]
@mrgrain I'm thinking about starting this and/or a Sveltekit project. Has anyone started one? If not, should I open a PR here for one or should that be a separate repo/package for now?

**Reply by mrgrain** [reply]
The preference are separate packages for now.

**Reply by automartin5000** [reply]
I get that for Sveltekit, but I feel like Vite could be a built-in component 😅

**Comment by kar-yeow** [comment]
Please add support for awscdk-app-dotnet.

**Reply by mrgrain** [reply]
Great suggestion! We also still need to publish projen for dotnet 🙈

**Comment by davidtwomey** [comment]
Hi @mrgrain,

New to projen, but looks very impressive.

We have a use-case for java, but with gradle kotlin DSL. 
I would be happy to attempt a PR if you could give me some pointers to get started.
I am unsure whether to extend the existing JavaProject, start from scratch, or create an alternative of the `pom.ts` (i.e. `gradle-kotlin.ts`

**Reply by mrgrain** [reply]
Glad you like it! My Java knowledge is limited, so I can only go off generalities here. 

To my understanding, Gradle is a replacement for Maven. I.e. they are mutually exclusive to each other(*). In that case an alternative Component to `pom.ts` would be the right approach. You might also find that the assumptions that Maven/POM is used, is (incorrectly) not limited to the POM component and you will have to untangle some of these assumptions. Finally you've also mentioned Kotlin and that might require further changes that will eventually deviate from the standard java project too much. 

(*) or at least in most real world setups they are

---

As project maintainers, we are currently gravitating towards having more project types separately maintained in external repos. I'd recommend you start with that and see how far you get. We can always link your project type from here and later talk about integrating it into core.

**Reply by davidtwomey** [reply]
Yes, typically they are mutually exclusive. I'll take a look at `pom.ts` and what is required to untangle it to be used more generically within the JavaProject.
Very briefly, maven uses `xml` for configuration, whereas gradle uses a different domain specific language (which come in two different forms: `groovy` and `kotlin`). So for my use-case I am referring to a **java** project which uses `gradle (kotlin DSL)`. 
I think (?) a full **kotlin** project would deviate far enough from the JavaProject to warrant a completely independent project type.

Anyway, thanks for your fast response! I'll have a go externally then. Is there a list of known external repos anywhere? might be some interesting examples I can look at

**Reply by mrgrain** [reply]
https://github.com/projen/awesome

**Comment by climbertjh** [comment]
I haven't looked super hard yet, but I'd like to see a couple examples of how to set up mono-repos which include sub-projects for:
- front-end (vite/React)
- back-end (Node/Express)
- infrastructure (CDK app)
- common (packages common to front-end/back-end, e.g. zod objects)

**Reply by mrgrain** [reply]
CDK app projects already exist.

**Reply by climbertjh** [reply]
I'll instantiate one of these to see if it creates something like what I'm thinking of.  Thanks.
